server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,403 @@
+                 * TODO : Refactor FileInfo and OnDemandBlockSnapshotIndexInput to more generic names as they are not Remote Snapshot specific
+                 */
+                BlobStoreIndexShardSnapshot.FileInfo fileInfo = new BlobStoreIndexShardSnapshot.FileInfo(
+                    name,
+                    new StoreFileMetadata(name, uploadedSegmentMetadata.getLength(), uploadedSegmentMetadata.getChecksum(), Version.LATEST),
+                    null
+                );
+                return new OnDemandBlockSnapshotIndexInput(fileInfo, localDirectory, transferManager);
+            }
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Closes the directory
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void close() throws IOException {
+        writeLock.lock();
+        try {
+            Arrays.stream(localDirectory.listAll()).forEach(f -> {
+                logger.trace("Removing file from cache {}", f);
+                fileCache.remove(localDirectory.getDirectory().resolve(f));
+            });
+            localDirectory.close();
+            remoteDirectory.close();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Returns a set of files currently pending deletion in this directory.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public Set<String> getPendingDeletions() throws IOException {
+        writeLock.lock();
+        try {
+            return localDirectory.getPendingDeletions();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Function to perform operations once files have been uploaded to Remote Store
+     * Currently deleting the local files here, as once uploaded to Remote, local files become eligible for eviction from FileCache
+     * @param files : recent files which have been successfully uploaded to Remote Store
+     * @throws IOException in case of I/O error
+     */
+    public void afterSyncToRemote(Collection<String> files) throws IOException {
+        logger.trace("afterSyncToRemote called for {}", files);
+        if (remoteDirectory == null) {
+            logger.trace("afterSyncToRemote called even though remote directory is not set");
+            return;
+        }
+        for (String fileName : files) {
+            writeLock.lock();
+            try {
+                /**
+                 * TODO - Unpin the files here from FileCache so that they become eligible for eviction, once pinning/unpinning support is added in FileCache
+                 * Uncomment the below commented line(to remove the file from cache once uploaded) to test block based functionality
+                 */
+                logger.trace("File {} uploaded to Remote Store and now can be eligible for eviction in FileCache", fileName);
+                // fileCache.remove(localDirectory.getDirectory().resolve(fileName));
```

Comment:
Done that as a temp measure in the latest revision. Will add file pinning in File Cache and then unpin the file from here so that it becomes eligible for eviction